File: backend/modules/produccion/services/trabajo_enlace.py
"""Validación del enlace opcional de trabajos; nunca se consulta la URL."""
from urllib.parse import urlsplit
import re
from ipaddress import IPv6Address

from modules.shared.exceptions import ValidationError

MAX_ENLACE = 2048


def validar_enlace(valor):
    if valor is None:
        return None
    mensaje = "Ingrese un enlace HTTP o HTTPS válido, de hasta 2048 caracteres."
    def rechazar():
        raise ValidationError(mensaje, details={"fields": {"enlace": mensaje}})
    if not isinstance(valor, str):
        rechazar()
    valor = valor.strip()
    if not valor:
        return None
    if len(valor) > MAX_ENLACE or re.search(r"[\s\x00-\x1f\x7f\\]", valor):
        rechazar()
    try:
        partes = urlsplit(valor)
        if partes.scheme.lower() not in ("http", "https") or not partes.hostname or partes.username is not None or partes.password is not None:
            rechazar()
        host = partes.hostname
        if ":" in host:
            IPv6Address(host)
        elif not re.fullmatch(r"[A-Za-z0-9.-]+", host.encode("idna").decode("ascii")):
            rechazar()
        partes.port  # Valida también puertos inválidos.
    except (ValueError, UnicodeError):
        rechazar()
    return valor

File: frontend/src/modules/produccion/utils/trabajoEnlace.ts
export const MAX_ENLACE = 2048;
export function errorEnlace(valor: string): string | undefined {
  const enlace = valor.trim();
  if (!enlace) return undefined;
  const mensaje = "Ingrese un enlace HTTP o HTTPS válido, de hasta 2048 caracteres.";
  if (/^https?:\/\/[^/?#]*@/i.test(enlace)) return mensaje;
  if (enlace.length > MAX_ENLACE || /[\s\x00-\x1f\x7f\\]/.test(enlace) || !/^https?:\/\//i.test(enlace)) return mensaje;
  try {
    const url = new URL(enlace);
    if (!url.hostname.startsWith("[") && !/^[A-Za-z0-9.-]+$/.test(url.hostname)) return mensaje;
    if (!["http:", "https:"].includes(url.protocol) || !url.hostname || url.username || url.password) return mensaje;
  } catch { return mensaje; }
  return undefined;
}
export function enlaceSeguro(valor?: string | null): string | null {
  return valor?.trim() && !errorEnlace(valor) ? valor.trim() : null;
}
